Quality checks on CI with LLVM/Clang tools

CarlosNihelton opened the pull request:
Hi @didrocks!

Here we go again.
As promissed in #66 I did the best I could to find the balance between the upstream original source files and capture it into the form of a `.clang-format` file. While I benefit from the help of ClangPowerTools Format Style Detector (I highly recommend it btw), I needed to loop in try-and-error for a while doing fine grained adjustments to the config and yet I couldn't get to the perfection, as the sources are not formatted consistently.

So I had to establish a couple of criteria:
- **Desconsider WslApiLoader.h and WslApiLoader.cpp**. Those files have very long lines and are far more divergent to the rest of the sources.
- **Avoid too long lines**. There are up to 147 chars long lines in the original sources (`DistributionInfo.cpp` line 48). That should not be our base line. There are lines in other files broken at 80th-90th column, and it is very common to find lines close to 100.
- **Purge tabs from our sources**. `OOBE.cpp` will be completely rewritten mainly due tabs.
- **Preserve what seems consistent across the original sources**. Brace breaks at function definitions and namespaces seem pretty common. That's not my preferred style, but it won't hurt and we'll benefit from the consistency.

## A theoretical exercise:

If we were to reformat the original sources per current clang format definition, it would look like this:

> Most consistent files: (`Distr*.{cpp,h}` and `Helpers.{cpp,h}`) -- notice how the headers wouldn't change.

```diff
diff --git a/DistroLauncher/DistributionInfo.cpp b/DistroLauncher/DistributionInfo.cpp
index 3367094..34c7fae 100644
--- a/DistroLauncher/DistributionInfo.cpp
+++ b/DistroLauncher/DistributionInfo.cpp
@@ -45,7 +45,8 @@ ULONG DistributionInfo::QueryUid(std::wstring_view userName)
         command += userName;
         int returnValue = 0;
         HANDLE child;
-        HRESULT hr = g_wslApi.WslLaunch(command.c_str(), true, GetStdHandle(STD_INPUT_HANDLE), writePipe, GetStdHandle(STD_ERROR_HANDLE), &child);
+        HRESULT hr = g_wslApi.WslLaunch(command.c_str(), true, GetStdHandle(STD_INPUT_HANDLE),
+                                        writePipe, GetStdHandle(STD_ERROR_HANDLE), &child);
         if (SUCCEEDED(hr)) {
             // Wait for the child to exit and ensure process exited successfully.
             WaitForSingleObject(child, INFINITE);
@@ -65,7 +66,7 @@ ULONG DistributionInfo::QueryUid(std::wstring_view userName)
                     try {
                         uid = std::stoul(buffer, nullptr, 10);
 
-                    } catch( ... ) { }
+                    } catch (...) { }
                 }
             }
         }
diff --git a/DistroLauncher/DistroLauncher.cpp b/DistroLauncher/DistroLauncher.cpp
index 83828f2..7c57f84 100644
--- a/DistroLauncher/DistroLauncher.cpp
+++ b/DistroLauncher/DistroLauncher.cpp
@@ -5,7 +5,7 @@
 
 #include "stdafx.h"
 
-// Commandline arguments: 
+// Commandline arguments:
 #define ARG_CONFIG              L"config"
 #define ARG_CONFIG_DEFAULT_USER L"--default-user"
 #define ARG_INSTALL             L"install"
@@ -29,7 +29,8 @@ HRESULT InstallDistribution(bool createUser)
         return hr;
     }
 
-    // Delete /etc/resolv.conf to allow WSL to generate a version based on Windows networking information.
+    // Delete /etc/resolv.conf to allow WSL to generate a version based on Windows networking
+    // information.
     DWORD exitCode;
     hr = g_wslApi.WslLaunchInteractive(L"rm /etc/resolv.conf", true, &exitCode);
     if (FAILED(hr)) {
@@ -38,7 +39,7 @@ HRESULT InstallDistribution(bool createUser)
 
     // Create a user account.
     if (createUser) {
-        if (DistributionInfo::isOOBEAvailable()){
+        if (DistributionInfo::isOOBEAvailable()) {
             return DistributionInfo::OOBESetup();
         }
         Helpers::PrintMessage(MSG_CREATE_USER_PROMPT);
@@ -103,7 +104,8 @@ int wmain(int argc, wchar_t const *argv[])
     if (!g_wslApi.WslIsDistributionRegistered()) {
 
         // If the "--root" option is specified, do not create a user account.
-        bool useRoot = ((installOnly) && (arguments.size() > 1) && (arguments[1] == ARG_INSTALL_ROOT));
+        bool useRoot =
+          ((installOnly) && (arguments.size() > 1) && (arguments[1] == ARG_INSTALL_ROOT));
         hr = InstallDistribution(!useRoot);
         if (FAILED(hr)) {
             if (hr == HRESULT_FROM_WIN32(ERROR_ALREADY_EXISTS)) {
@@ -128,8 +130,7 @@ int wmain(int argc, wchar_t const *argv[])
                 Helpers::PromptForInput();
             }
 
-        } else if ((arguments[0] == ARG_RUN) ||
-                   (arguments[0] == ARG_RUN_C)) {
+        } else if ((arguments[0] == ARG_RUN) || (arguments[0] == ARG_RUN_C)) {
 
             std::wstring command;
             for (size_t index = 1; index < arguments.size(); index += 1) {
diff --git a/DistroLauncher/Helpers.cpp b/DistroLauncher/Helpers.cpp
index 524e8b6..e9809a5 100644
--- a/DistroLauncher/Helpers.cpp
+++ b/DistroLauncher/Helpers.cpp
@@ -5,7 +5,8 @@
 
 #include "stdafx.h"
 
-namespace {
+namespace
+{
     HRESULT FormatMessageHelperVa(DWORD messageId, va_list vaList, std::wstring* message);
     HRESULT PrintMessageVa(DWORD messageId, va_list vaList);
 }
@@ -32,7 +33,7 @@ std::wstring Helpers::GetUserInput(DWORD promptMsg, DWORD maxCharacters)
 
 void Helpers::PrintErrorMessage(HRESULT error)
 {
-    PWSTR buffer = nullptr; 
+    PWSTR buffer = nullptr;
     ::FormatMessageW(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_ALLOCATE_BUFFER,
                      nullptr,
                      error,
@@ -65,17 +66,19 @@ void Helpers::PromptForInput()
     return;
 }
 
-namespace {
+namespace
+{
     HRESULT FormatMessageHelperVa(DWORD messageId, va_list vaList, std::wstring* message)
     {
-        PWSTR buffer = nullptr; 
-        DWORD written = ::FormatMessageW(FORMAT_MESSAGE_FROM_HMODULE | FORMAT_MESSAGE_ALLOCATE_BUFFER,
-                                         nullptr,
-                                         messageId,
-                                         0,
-                                         (PWSTR)&buffer,
-                                         10,
-                                         &vaList);
+        PWSTR buffer = nullptr;
+        DWORD written =
+          ::FormatMessageW(FORMAT_MESSAGE_FROM_HMODULE | FORMAT_MESSAGE_ALLOCATE_BUFFER,
+                           nullptr,
+                           messageId,
+                           0,
+                           (PWSTR)&buffer,
+                           10,
+                           &vaList);
         *message = buffer;
         if (buffer != nullptr) {
             HeapFree(GetProcessHeap(), 0, buffer);

``` 
> The most complicated ones (`WslApiLoader.{cpp,h}`)

```diff
diff --git a/DistroLauncher/WslApiLoader.cpp b/DistroLauncher/WslApiLoader.cpp
index 53fdf83..1032fef 100644
--- a/DistroLauncher/WslApiLoader.cpp
+++ b/DistroLauncher/WslApiLoader.cpp
@@ -11,11 +11,16 @@ WslApiLoader::WslApiLoader(const std::wstring& distributionName) :
 {
     _wslApiDll = LoadLibraryEx(L"wslapi.dll", nullptr, LOAD_LIBRARY_SEARCH_SYSTEM32);
     if (_wslApiDll != nullptr) {
-        _isDistributionRegistered = (WSL_IS_DISTRIBUTION_REGISTERED)GetProcAddress(_wslApiDll, "WslIsDistributionRegistered");
-        _registerDistribution = (WSL_REGISTER_DISTRIBUTION)GetProcAddress(_wslApiDll, "WslRegisterDistribution");
-        _configureDistribution = (WSL_CONFIGURE_DISTRIBUTION)GetProcAddress(_wslApiDll, "WslConfigureDistribution");
-        _getDistributionConfiguration = (WSL_GET_DISTRIBUTION_CONFIGURATION)GetProcAddress(_wslApiDll, "WslGetDistributionConfiguration");
-        _launchInteractive = (WSL_LAUNCH_INTERACTIVE)GetProcAddress(_wslApiDll, "WslLaunchInteractive");
+        _isDistributionRegistered =
+          (WSL_IS_DISTRIBUTION_REGISTERED)GetProcAddress(_wslApiDll, "WslIsDistributionRegistered");
+        _registerDistribution =
+          (WSL_REGISTER_DISTRIBUTION)GetProcAddress(_wslApiDll, "WslRegisterDistribution");
+        _configureDistribution =
+          (WSL_CONFIGURE_DISTRIBUTION)GetProcAddress(_wslApiDll, "WslConfigureDistribution");
+        _getDistributionConfiguration = (WSL_GET_DISTRIBUTION_CONFIGURATION)GetProcAddress(
+          _wslApiDll, "WslGetDistributionConfiguration");
+        _launchInteractive =
+          (WSL_LAUNCH_INTERACTIVE)GetProcAddress(_wslApiDll, "WslLaunchInteractive");
         _launch = (WSL_LAUNCH)GetProcAddress(_wslApiDll, "WslLaunch");
     }
 }
@@ -29,12 +34,9 @@ WslApiLoader::~WslApiLoader()
 
 BOOL WslApiLoader::WslIsOptionalComponentInstalled()
 {
-    return ((_wslApiDll != nullptr) && 
-            (_isDistributionRegistered != nullptr) &&
-            (_registerDistribution != nullptr) &&
-            (_configureDistribution != nullptr) &&
-            (_getDistributionConfiguration != nullptr) &&
-            (_launchInteractive != nullptr) &&
+    return ((_wslApiDll != nullptr) && (_isDistributionRegistered != nullptr) &&
+            (_registerDistribution != nullptr) && (_configureDistribution != nullptr) &&
+            (_getDistributionConfiguration != nullptr) && (_launchInteractive != nullptr) &&
             (_launch != nullptr));
 }
 
@@ -53,9 +55,11 @@ HRESULT WslApiLoader::WslRegisterDistribution()
     return hr;
 }
 
-HRESULT WslApiLoader::WslConfigureDistribution(ULONG defaultUID, WSL_DISTRIBUTION_FLAGS wslDistributionFlags)
+HRESULT WslApiLoader::WslConfigureDistribution(ULONG defaultUID,
+                                               WSL_DISTRIBUTION_FLAGS wslDistributionFlags)
 {
-    HRESULT hr = _configureDistribution(_distributionName.c_str(), defaultUID, wslDistributionFlags);
+    HRESULT hr =
+      _configureDistribution(_distributionName.c_str(), defaultUID, wslDistributionFlags);
     if (FAILED(hr)) {
         Helpers::PrintMessage(MSG_WSL_CONFIGURE_DISTRIBUTION_FAILED, hr);
     }
@@ -65,7 +69,7 @@ HRESULT WslApiLoader::WslConfigureDistribution(ULONG defaultUID, WSL_DISTRIBUTIO
 
 HRESULT WslApiLoader::WslGetDistributionConfiguration(ULONG* distributionVersion,
                                                       ULONG* defaultUID,
-                                                      WSL_DISTRIBUTION_FLAGS* wslDistributionFlags, 
+                                                      WSL_DISTRIBUTION_FLAGS* wslDistributionFlags,
                                                       PSTR** defaultEnvironmentVariables,
                                                       ULONG* defaultEnvironmentVariableCount)
 {
@@ -77,9 +81,12 @@ HRESULT WslApiLoader::WslGetDistributionConfiguration(ULONG* distributionVersion
                                          defaultEnvironmentVariableCount);
 }
 
-HRESULT WslApiLoader::WslLaunchInteractive(PCWSTR command, BOOL useCurrentWorkingDirectory, DWORD *exitCode)
+HRESULT WslApiLoader::WslLaunchInteractive(PCWSTR command,
+                                           BOOL useCurrentWorkingDirectory,
+                                           DWORD* exitCode)
 {
-    HRESULT hr = _launchInteractive(_distributionName.c_str(), command, useCurrentWorkingDirectory, exitCode);
+    HRESULT hr =
+      _launchInteractive(_distributionName.c_str(), command, useCurrentWorkingDirectory, exitCode);
     if (FAILED(hr)) {
         Helpers::PrintMessage(MSG_WSL_LAUNCH_INTERACTIVE_FAILED, command, hr);
     }
@@ -87,9 +94,20 @@ HRESULT WslApiLoader::WslLaunchInteractive(PCWSTR command, BOOL useCurrentWorkin
     return hr;
 }
 
-HRESULT WslApiLoader::WslLaunch(PCWSTR command, BOOL useCurrentWorkingDirectory, HANDLE stdIn, HANDLE stdOut, HANDLE stdErr, HANDLE *process)
+HRESULT WslApiLoader::WslLaunch(PCWSTR command,
+                                BOOL useCurrentWorkingDirectory,
+                                HANDLE stdIn,
+                                HANDLE stdOut,
+                                HANDLE stdErr,
+                                HANDLE* process)
 {
-    HRESULT hr = _launch(_distributionName.c_str(), command, useCurrentWorkingDirectory, stdIn, stdOut, stdErr, process);
+    HRESULT hr = _launch(_distributionName.c_str(),
+                         command,
+                         useCurrentWorkingDirectory,
+                         stdIn,
+                         stdOut,
+                         stdErr,
+                         process);
     if (FAILED(hr)) {
         Helpers::PrintMessage(MSG_WSL_LAUNCH_FAILED, command, hr);
     }
diff --git a/DistroLauncher/WslApiLoader.h b/DistroLauncher/WslApiLoader.h
index ef89fe6..b513ad2 100644
--- a/DistroLauncher/WslApiLoader.h
+++ b/DistroLauncher/WslApiLoader.h
@@ -11,12 +11,17 @@
 #define ERROR_LINUX_SUBSYSTEM_NOT_PRESENT 414L
 #endif // !ERROR_LINUX_SUBSYSTEM_NOT_PRESENT
 
-typedef BOOL    (STDAPICALLTYPE* WSL_IS_DISTRIBUTION_REGISTERED)(PCWSTR);
-typedef HRESULT (STDAPICALLTYPE* WSL_REGISTER_DISTRIBUTION)(PCWSTR, PCWSTR);
-typedef HRESULT (STDAPICALLTYPE* WSL_CONFIGURE_DISTRIBUTION)(PCWSTR, ULONG, WSL_DISTRIBUTION_FLAGS);
-typedef HRESULT (STDAPICALLTYPE* WSL_GET_DISTRIBUTION_CONFIGURATION)(PCWSTR, ULONG *, ULONG *, WSL_DISTRIBUTION_FLAGS *, PSTR **, ULONG *);
-typedef HRESULT (STDAPICALLTYPE* WSL_LAUNCH_INTERACTIVE)(PCWSTR, PCWSTR, BOOL, DWORD *);
-typedef HRESULT (STDAPICALLTYPE* WSL_LAUNCH)(PCWSTR, PCWSTR, BOOL, HANDLE, HANDLE, HANDLE, HANDLE *);
+typedef BOOL(STDAPICALLTYPE* WSL_IS_DISTRIBUTION_REGISTERED)(PCWSTR);
+typedef HRESULT(STDAPICALLTYPE* WSL_REGISTER_DISTRIBUTION)(PCWSTR, PCWSTR);
+typedef HRESULT(STDAPICALLTYPE* WSL_CONFIGURE_DISTRIBUTION)(PCWSTR, ULONG, WSL_DISTRIBUTION_FLAGS);
+typedef HRESULT(STDAPICALLTYPE* WSL_GET_DISTRIBUTION_CONFIGURATION)(PCWSTR,
+                                                                    ULONG*,
+                                                                    ULONG*,
+                                                                    WSL_DISTRIBUTION_FLAGS*,
+                                                                    PSTR**,
+                                                                    ULONG*);
+typedef HRESULT(STDAPICALLTYPE* WSL_LAUNCH_INTERACTIVE)(PCWSTR, PCWSTR, BOOL, DWORD*);
+typedef HRESULT(STDAPICALLTYPE* WSL_LAUNCH)(PCWSTR, PCWSTR, BOOL, HANDLE, HANDLE, HANDLE, HANDLE*);
 
 class WslApiLoader
 {
@@ -30,26 +35,22 @@ class WslApiLoader
 
     HRESULT WslRegisterDistribution();
 
-    HRESULT WslConfigureDistribution(ULONG defaultUID,
-                                     WSL_DISTRIBUTION_FLAGS wslDistributionFlags);
+    HRESULT WslConfigureDistribution(ULONG defaultUID, WSL_DISTRIBUTION_FLAGS wslDistributionFlags);
 
     HRESULT WslGetDistributionConfiguration(ULONG* distributionVersion,
                                             ULONG* defaultUID,
                                             WSL_DISTRIBUTION_FLAGS* wslDistributionFlags,
                                             PSTR** defaultEnvironmentVariables,
-                                            ULONG* defaultEnvironmentVariableCount
-                                            );
+                                            ULONG* defaultEnvironmentVariableCount);
 
-    HRESULT WslLaunchInteractive(PCWSTR command,
-                                 BOOL useCurrentWorkingDirectory,
-                                 DWORD *exitCode);
+    HRESULT WslLaunchInteractive(PCWSTR command, BOOL useCurrentWorkingDirectory, DWORD* exitCode);
 
     HRESULT WslLaunch(PCWSTR command,
                       BOOL useCurrentWorkingDirectory,
                       HANDLE stdIn,
                       HANDLE stdOut,
                       HANDLE stdErr,
-                      HANDLE *process);
+                      HANDLE* process);
 
   private:
     std::wstring _distributionName;

```
Of course that's only an exercise. The changes I am applying for review are only in our files.

If the proposed style is good enough for you, let's get this merged so, as long as we let the tools to their job, we should never talk about formatting issues in C++ sources from now on.

didrocks commented:
> Avoid too long lines. There are up to 147 chars long lines in the original sources (DistributionInfo.cpp line 48). That should not be our base line. There are lines in other files broken at 80th-90th column, and it is very common to find lines close to 100.

Can we make it 120 characters? I always find 80 or 100 too narrow and try to stick to that (which is compatible with most screen sizes and IDEs with a navigation bar).

The rest of the style guide sounds sensible to me.

Two things though:
* Is there any way to define which files to ignore in `.clang-format`? That way, no question on upstream files which may be formatted on saving by mistake without noticing.
* Can we have that in CI on pull request as part of a linter check? This is what we mostly do in other projects.

CarlosNihelton commented:
> > Avoid too long lines. There are up to 147 chars long lines in the original sources (DistributionInfo.cpp line 48). That should not be our base line. There are lines in other files broken at 80th-90th column, and it is very common to find lines close to 100.
> 
> Can we make it 120 characters? I always find 80 or 100 too narrow and try to stick to that (which is compatible with most screen sizes and IDEs with a navigation bar).

Sure we can.
 
> Two things though:
> 
> * Is there any way to define which files to ignore in `.clang-format`? That way, no question on upstream files which may be formatted on saving by mistake without noticing.

Not exactly in .clang-format itself, but there are a number of ways to doing so. Upstream files could be formatted by calng-formatin three ways (as I can think of now):
- Intentionally from the editor, by opening the file and commanding the editor to format the open file.
- Unintentionally by saving the document (if format on save is flagged in the editor)
- via CLI (explicitly or implicitly -- like commanding `clang-format -i *.cpp *h`)

I don't see how to prevent those situations with clang-format, but we can:
- Add a check in the CI to run clang-format explicitly on those files, printing the output to stdout instead of in-place editing the files, and testing for empty input (that would mean the files were formatted and the CI would show same warning or error). IT's somewhat fragile, but it would help.
- Add a (custom maybe) wrapper script to the workflow, to be invoked both by CI and by developers during normal workflow to format all source files, except the ignored ones. [Run-clang-format](https://github.com/Sarcasm/run-clang-format/blob/master/run-clang-format.py) for instance supports a `.clang-format-ignore` file with a syntax very similar to `.gitignore` 

> * Can we have that in CI on pull request as part of a linter check? This is what we mostly do in other projects.
Sure we can! And we can do it and two ways (like the security modules 😅 ):
- Enforcing: similar to what you did with `refresh releases and assets`
- Complain: detect unformatted sources, blame the requester to fix formatting and refuse to merge.

Whatever direction we choose, I think I'd go with a solution similar to the python script mentioned above for the CI.

Could you give me a good example of CI workflow I should follow on building this?

CarlosNihelton commented:
There are also very interesting Github Actions we can use, like:
https://github.com/marketplace/actions/c-c-lint-action
and
https://github.com/marketplace/actions/clang-format-check

CarlosNihelton commented:
> There are also very interesting Github Actions we can use, like: https://github.com/marketplace/actions/c-c-lint-action and https://github.com/marketplace/actions/clang-format-check

The first will match nicely with the [Annotations action](https://github.com/marketplace/actions/annotations-action) to show the output of the linting process right into the code. That seems really interesting. What do you think about it?

didrocks commented:
Ack on not being able to do this directly from the .clang-format file, this is a little bit sad IMHO :)

On adys, this is part of this step of the Code quality job: https://github.com/ubuntu/adsys/blob/main/.github/workflows/qa.yaml#L24. Before using golangci-lint, we used to do this:
* get out of the index some files (which can be "blacklist" files from original upstream
* `go fmt`
* `check if there is anything in `git diff`

Check one of the later version on lines https://github.com/ubuntu/adsys/blob/20e53162ca6fae151fa48d52f7152c3355a9a827/.github/workflows/qa.yaml#L28 and https://github.com/ubuntu/adsys/blob/20e53162ca6fae151fa48d52f7152c3355a9a827/.github/workflows/qa.yaml#L52.

This was before Github action introduced the annotation helpers to show the annotation in the diff, which is really interested.

However, we try to select third party actions (like those not provided by trusted companies) carrefully: your action has access to a lot of things (like secrets in the case of WSL) and we don’t want them to tamper our keys. I’m unsure about using the first action for this reason.

But yeah, CI on PR as a separate job is the way to go then IMHO (with ideally, annotations)

CarlosNihelton commented:
Let's start again.

This changed so much that I preferred to force-push and start fresh. The Quality CI workflow is organized in such a way that:

- Review comments are generated from clang-format (the formatter) and clang-tidy (the linter).
- Only code modified in the pull request triggering the workflow will be commented. Technical debts will be addressed by our proactiviy or by warnings generated when the affected files get modified by future pull requests.
- Comments generated come with fix suggestion when possible. Format issues should always generate suggestions, but linting issues most certainly not.
- Some preparation steps very specific to this project were required to leverage the linter because msbuild doesn't export compilation database in the format expected by clang-tidy.
- Code was subdivided in such a way that we can, at some other moment, extract pieces of it and reuse in other projects and other languages and tools.

---

The suggestion of 120 columns was adopted.

CarlosNihelton commented:
I have one more push to do, but I'll wait for PR #74 to get merged, because if the proposed auto tests get merged, I'll make this QA job run them. The branch is ready to be pushed, but I'll do it just yet otherwise the QA job will fail due lack of the tests solution. Also, in that push I'll get the dependency management and the hacks imposed by it completely removed from this QA job. I'll probably need to do some rebase locally and force pushing. We'll see.

CarlosNihelton commented:
# Recap!

The last push was a big remake of this pull request. To recap, let's revisit what were the key objectives:

- [x] Establish a CI workflow to check pull requests for quality.
- [x] Ensure standard formatting. Done by added a project level `.clang-format` file. Visual Studio respects it, but also the CI workflow enforces it.
- [x] Establish a linting process. Done by setting up msbuild to run clang-tidy to lint the sources. Checks are defined in `.clang-tidy` file.
- [x] Ignore upstream original sources to avoid creating unnecessary diffs. Done by creating a `.clang-ignore` file (similar in syntax) to .gitignore and custom runners to ensure clang-format and clang-tidy will respect the definitions of that file.
- [x] Automatically comment the pull request when formatting issues are detected. Done by creating comments with suggestions from the possible diff generated when misformatted files are detected in the pull request. Files not touched in the PR being checked are not considered.
- [x] Automatically comment the pull request when linting issues are detected. Done by making clang-tidy to export a `fixes.yaml` file with the issues detected, parsing that file and posting comments with suggestions when clang-tidy proposes fixes for the reported issues. That required some customisation of the msbuild workflow. See the contents of `msbuild/` inside the workflows directory.
- [x] Avoid repeating comments when PR is updated with new commits. 
- [x] Automatic testing. Done by creating another solution in a separate folder (DistroLauncher-Tests) with support for mocking (using Google Testing Framework). It's way far from being perfect, but is a starting point to enables us validating our theories and assumptions. To demonstrate it I added testing for the OOBE exit status file parser, which is a little more evolving and could easily be got wrong. The tests I implemented validated that I made what I intended.
- [x] Remove the need for vcpkg in the CI. Since we are not fetching dependencies anymore in this project, dependency management is unnecessary and was removed from the build workflow.

Notice that, by design, the workflow should not fail when formatting or linting issues are detected. Instead, comments are generated. But if the tests fail to compile or to pass or yet if clang-tidy detects relevant enough issues, it triggers "compilation error" from the perspective of msbuild and thus the workflow fast-fails, before being able to generate pull request review comments. 

## How to test it

The latest demonstration of the features of this workflow is available in my personal fork, [pull request 7](https://github.com/CarlosNihelton/WSL/pull/7). The only difference between that pull request and this one is one commit in which I reformatted one file to cause the workflow to be triggered. If you feel the need to test it in this repository, it's just a matter of creating a new pull request with changes in one of our C++ source files with know formatting mismatches or lines not complaint with clang-tidy checks expectations.

didrocks commented on DistroLauncher-Tests/DistroLauncher-Tests/main.cpp:
```diff
@@ -0,0 +1,10 @@
+#include "stdafx.h"
+#include <cstdio>
+#include "gtest/gtest.h"
```
How is gtest pulled? We just expect to have the header installed on the machine (which is ok IMHO) and the `GTest/` subfolder will just compile/link against it?

didrocks replied:
Ack, thanks for the explanation! :)

didrocks reviewed (changes requested):
Phew, finally done! I think few changes could be done, but I really prefer this than relying on deps + cmake transformation.
Of course, due to the length of changes, I might have missed some logic issue, but the linked PR example, your detailed summary (thank you!) and the rest makes me confident in this one :)

If you address the comments (or find them utterly stupids), you can merge it afterwards without having to wait on me again. Thanks!

CarlosNihelton commented on .github/workflows/clang_lint/inline_comments/post_pr_review_comments.py:
```diff
@@ -0,0 +1,226 @@
+                    int(change.split(",")[0]) + int(change.split(",")[1]),
+                )
+            )
+            for change in lines_and_changes
```
😎

CarlosNihelton commented on .github/workflows/qa.yaml:
```diff
@@ -0,0 +1,39 @@
+          msbuild -p:Platform=x64 -p:Configuration=Debug .\DistroLauncher-Tests\DistroLauncher-Tests.sln
+          .\DistroLauncher-Tests\x64\Debug\DistroLauncher-Tests.exe
+      # Produces the fixes.yaml to be consumed by the commenter.
+      - name: Run Clang-Tidy.
```
I understand your point, but clang-tidy runs like a compiler, so having it running right after the tests enables reusing some compilation artifacts, like the precompiled headers and the `messages.h` file generated from the resource file `messages.mc`. If we split that in two workflows this reuse will not be possible, which results in making some redundant work. Not much, but still redundant. Do you think the clarity gained by splitting that up worths this somewhat redundancy?

CarlosNihelton replied:
Done. I made running the tests on its own worflow. We should be good to go.

didrocks replied:
Thanks for considering, let’s merge it now!

CarlosNihelton commented on .github/workflows/clang_lint/__init__.py:
```diff
@@ -0,0 +1,18 @@
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#
```
I think it's easier to preserve the consistency this way than expecting the excited developer to remember to add the copyright when, for any reason in the future, he decides that this file should not be empty anymore.

didrocks replied:
Sounds good to me :)

CarlosNihelton commented on .github/workflows/clang_lint/clang_lint/custom_clang_format_commenter.py:
```diff
@@ -0,0 +1,62 @@
+    pr_review = diff_to_review.diff_to_review(repository_root, git_diff,
+                                              body, inline_msg, sources)
+    # Reset repo, otherwise the next one in the pipeline will work on a
+    #  modified tree different from what GitHub displays on browser.
```
😅

CarlosNihelton commented on .github/workflows/clang_lint/clang_lint/custom_clang_tidy_commenter.py:
```diff
@@ -0,0 +1,192 @@
+from clang_lint.inline_comments.markdown import markdown
+
+
+def parse_clang_tidy_fixes(fixes_path: str, repository_root: str):
```
Indeed. My bad. Fixed now.

CarlosNihelton commented on .github/workflows/clang_lint/diff_to_review/diff_to_review.py:
````diff
@@ -0,0 +1,60 @@
+            target_lines = [
+                tl.value for tl in hunk.target_lines()
+            ]
+            review_comment_body = inline_msg + "\n\n```suggestion\n" + \
````
Sometimes I agree, sometimes I just forget... Addressed.

didrocks replied:
next time, promise, I will make a comment without 2 typos in the first 10 words…

CarlosNihelton commented on .github/workflows/clang_lint/clang_lint/custom_clang_format_commenter.py:
```diff
@@ -0,0 +1,62 @@
+        print("Failed to run clang-format.")
+        return None
+
+    repo = git.Repo(repository_root)
```
Good point. This piece of the workflow (as is) totally depends on Git. Thus wrapping in a try-except clause and failing fast makes sense.
It would be possible to make a copy of the work directory before running the formatter and running the Python equivalent to `git diff --no-index dirA dirB`, which obviously requires detecting first whether we are in a git repository or not. Do you think it is worthy? For now I'm just pushing the try-except.

didrocks replied:
The try-except is good enough IMHO for now. At least, if we install it on a non git repo, we will see it quick and then, decide if the copy + manual diff is better or not. Thanks!

didrocks commented:
Just one remaining comment to be resolved (which could be a no-op) and then, good for merging!

didrocks commented:
I will let you merge :)
